src/fm.py

```diff
@@ -33,5 +33,172 @@ def main():
+        m = (lo + hi) // 2
+        if x[sa[m] + offset % len(x)] < a: # compare at column offset in sa
+            lo = m + 1
+        else:
+            hi = m
+    return lo
+
+def upper(a: str, x: str, sa: memoryview, lo: int, hi: int, offset: int) -> int:
+    """Finds the upper bound of `a` at `offset` in the block defined by `lo:hi`."""
+    return lower(chr(ord(a) + 1), x, sa, lo, hi, offset)
+
+def search(sa, pattern, genome):
+    lo, hi = 0, len(sa)
+    for offset, a in enumerate(pattern):
+        lo = lower(a, genome, sa, lo, hi, offset)
+        hi = upper(a, genome, sa, lo, hi, offset)
+    for sol in sa[lo:hi]:
+        yield sol+1
+    return
+
+def getSuffixes(x):
+    """
+    Gets all suffixes from a string
+    """
+    suffixes = [x[i:] for i in range(0, len(x))] 
+    # print("list of unordered suffixes: ",suffixes) 
+    return suffixes
+
+def radix_sort(lst: list[memoryview]):
+    # print("Radix sort", len(lst))
+    maximum = max(lst, key = len)
+
+    for place in range(len(maximum),0, -1):
+        lst = counting_sort(lst, place - 1)
+    
+    lst = [len(maximum) - len(suffix) for suffix in lst]
+    return lst
+            
+
+def counting_sort(lst: list[memoryview], place):
+    maximum = max(lst, key = len)
+    counts = dict.fromkeys(["$",*sorted(maximum)],[])
+    for string_index in range(len(lst)):
+        if place >= len(lst[string_index]):
+            counts["$"] = [*counts["$"], lst[string_index]]
+            
+        else:
+            counts[lst[string_index][place]] = [*counts[lst[string_index][place]], lst[string_index]]
+
+    ordered = []
+    for key in counts:
+        ordered += counts[key]
+    return ordered
+
+def build_rank_table(x, alphadic, bwt):
+    counts = {a: 0 for a in alphadic}
+    table = [[0 for _ in alphadic] for _ in range(0, len(bwt)+1)]
+
+    for i in range(1, len(bwt)+1):        
+        for j in range(0, len(alphadic)):
+            table[i][j] = table[i-1][j]
+        
+        bwtValue = bwt[i-1]
+        c = chr(x[bwtValue])
+
+        # Add the count
+        counts[c] += 1
+
+        index = alphadic.get(c)
+        table[i][index] += 1
+
+    return table, counts
+
+def getFirstIndexList(x, f, alphadic):
+    firstIndexList = {a: -1 for a in alphadic}
+    indexesFound = 0
+
+    for i, xIndex in enumerate(f):   
+        c = chr(x[xIndex])
+        if firstIndexList.get(c) < 0:
+            firstIndexList[c] = i
+            indexesFound += 1
+
+            if indexesFound >= len(alphadic):
+                return firstIndexList
+    
+    return firstIndexList
+
+
+def getrank(alphadic, index, c, rank_table):
+    return rank_table[index][alphadic.get(c)]
+
+def searchPattern(x, p, f, l, rank_table, firstIndexList, countTable, alphadic):
+    if p == "":
+        return
+    
+    if len(p) == 1:
+        firstMatch = firstIndexList.get(p[0])
+        for i in range(0, countTable.get(p[0])):
+            yield f[i + firstMatch]
+        return
```

Comment:
You don't actually need a special case here. The general case can handle `len(p) == 1`. (It also handles `len(p) == 0` if you do it right).